### ISSUE TITLE
Loss and accuracy calculation

### ISSUE THREAD

**Issue by JesperHenriksen** [sotte/pytorch_tutorial#10]
Hallo.

Great tutorial!
In your loss and accuracy calculations you divide the loss and correct with the length of the entire data set like this:

Train Loss:  running_loss / len(train_dl.dataset)
Train Acc:  correct / len(train_dl.dataset)

Shouldn't it be the loss and correct divided by the batch size? Then you get the accuracy of the current batch like this:

        ite = 0
        for i, (X, y) in enumerate(train_dl):
            X, y = X.to(DEVICE), y.to(DEVICE)

            optimizer.zero_grad()
            y_ = model(X)
            loss = criterion(y_, y)
            loss.backward()
            optimizer.step()

            # Statistics
            _, y_label_ = torch.max(y_, 1)
            correct += (y_label_ == y).sum().item()
            running_loss += loss.item() * X.shape[0]
            ite = i
        print(f"  Train Loss: {running_loss / ite+1}")
        print(f"  Train Acc:  {correct /  ite+1}")

Edit: Nevermind, i am an idiot. Because you go through x images per batch you divide with number of images and what i do is dividing the loss and correct with number of batches. I will never get the correct loss

Just close this issue

2nd Edit: Okay so i found this https://discuss.pytorch.org/t/interpreting-loss-value/17665/9 
They claim that the loss should be calculated in a different way. Who is right i cannot say.

**Comment by sotte** [comment]
Thank you for the kind words!

Just real quick: if you look at loss functions you'll often see the `reduction` argument, e.g. here for [cross entropy](https://pytorch.org/docs/stable/nn.html#crossentropyloss) it's "mean". So what you get when you do `loss = criterion(y_, y)` is the cross entropy loss averaged over the batch.
In theory batches can have different sizes (and the last batch normally is smaller). One should take this into account. 

As I said, writing a train loop and metrics is very error prone and it's not unlikely that I made a mistake (for educational purposes of course ;)) What notebook are you talking about exactly? I'll double check.

**Comment by JesperHenriksen** [comment]
Haha ofc it was for the educational purpose :D
Transfer learning in the training loop in the bottom. The guy from the link claims that this is correct:

"""
Here is what you need to do, define
num_images = 0

At each batch iteration do

num_images += inputs.size(0)

Note that inputs.size(0) would give the total images in each batch.
Finally

epoch_loss = running_loss/num_images
epoch_acc = running_acc/num_images

That would give you accurate metrics
"""
And have the same calculations as you have with running_loss += loss.item() * X.shape[0]

Edit: When i tried his version with my model it made a lot more sense the results i got.

**Comment by sotte** [comment]
Thanks so much for you input!
Of course I was wrong and I actually said so during the workshop: 
https://youtu.be/_H3aw6wkCv0?t=4495
One always gets the loop wrong ;D

I'll push a fix in a second. It should look something like this.
```python
    # Train
    model.train()  # IMPORTANT
    
    total_loss, n_correct, n_samples = 0.0, 0, 0
    for batch_i, (X, y) in enumerate(train_dl):
        X, y = X.to(DEVICE), y.to(DEVICE)
        ...
        
        # Statistics
        _, y_label_ = torch.max(y_, 1)
        n_correct += (y_label_ == y).sum().item()
        total_loss += loss.item() * X.shape[0]
        n_samples += X.shape[0]
    
    print(
        f"Epoch {epoch+1}/{N_EPOCHS} |"
        f"  train loss: {total_loss / n_samples:9.3f} |"
        f"  train acc:  {n_correct / n_samples * 100:9.3f}%"
    )
```
The loss must be multiplied by the number of samples. I keep track of the total number of samples in `n_samples`. When printing the stats I divide the `total_loss` by `n_samples` to get the average :)

**Comment by sotte** [comment]
Closed by 1f166a34328d564f5b65da50d49519037bc21b76
